server/src/test/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunnerTests.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+public class ThrottledTaskRunnerTests extends ESTestCase {
+    private ThreadPool threadPool;
+    private Executor executor;
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        threadPool = new TestThreadPool("test");
+        executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        TestThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+    }
+
+    static class TestTask implements Comparable<TestTask>, Runnable {
+
+        private final Runnable runnable;
+
+        TestTask(Runnable runnable) {
+            this.runnable = runnable;
+        }
+
+        @Override
+        public int compareTo(TestTask o) {
+            return randomIntBetween(0, 1000);
```

Comment:
I'd think this should be between (-1000, 1000).

But I can imagine `PrioritizedBlockingQueue` not working with such a compare. I think it expects objects to adhere to the `compare` contract, which requires stability and transitivity.

Can we assign a random number to the task and compare the tasks using that number instead?

Can we also validate that we execute tasks in order?

Reply:
done in 56619c0eb675ec0401f7ddf73b7a1950bf03677c
I added a new test for the ordering in 79331f810c3a4f21152d57ad3d99a0356fdbf5af